dev-docs/references/testing.md

```diff
@@ -0,0 +1,254 @@
+logic for those tests.
+
+By default, integration tests will do the following on a given cloud platform:
+ * Launch an instance running latest daily image of the target Ubuntu release
+ * Add the Ubuntu advantage client daily build PPA: [ppa:ua-client/daily](https://code.launchpad.net/~ua-client/+archive/ubuntu/daily)
+ * Install the appropriate ubuntu-advantage-tools and ubuntu-advantage-pro deb
+ * Run the integration tests on that instance.
+
+The testing can be overridden to run using a local copy of the ubuntu-advantage-client source code instead of the daily PPA by providing the following environment variable to the behave test runner:
+```UACLIENT_BEHAVE_BUILD_PR=1```
+
+> Note that, by default, we cache the source even when `UACLIENT_BEHAVE_BUILD_PR=1`. This means that if you change the python code locally and want to run the behave tests against your new version, you need to either delete the cache (`rm /tmp/uaclient_behave`) or also set `UACLIENT_BEHAVE_CACHE_SOURCE=0`.
+
+To run the tests, you can use `tox`:
+
+```shell
+tox -e behave-lxd-20.04
+```
+
+or, if you just want to run a specific file, or a test within a file:
+
+```shell
+tox -e behave-lxd-20.04 features/unattached_commands.feature
+tox -e behave-lxd-20.04 features/unattached_commands.feature:55
+```
+
+As can be seen, this will run behave tests only for release 20.04 (Focal Fossa). We are currently
+supporting 5 distinct releases:
+
+* 22.04 (Jammy Jellyfish)
+* 21.10 (Impish Indri)
+* 20.04 (Focal Fossa)
+* 18.04 (Bionic Beaver)
+* 16.04 (Xenial Xerus)
+
+Therefore, to change which release to run the behave tests against, just change the release version
+on the behave command.
+
+Furthermore, when developing/debugging a new scenario:
+
+ 1. Add a `@wip` tag decorator on the scenario
+ 2. To only run @wip scenarios run: `tox -e behave-lxd-20.04 -- -w`
+ 3. If you want to use a debugger:
+    1. Add ipdb to integration-requirements.txt
+    2. Add ipdb.set_trace() in the code block you wish to debug
+
+(If you're getting started with behave, we recommend at least reading
+through [the behave
+tutorial](https://behave.readthedocs.io/en/latest/tutorial.html) to get
+an idea of how it works, and how tests are written.)
+
+## Iterating Locally
+
+To make running the tests repeatedly less time-intensive, our behave
+testing setup has support for reusing images between runs via two
+configuration options (provided in environment variables),
+`UACLIENT_BEHAVE_IMAGE_CLEAN` and `UACLIENT_BEHAVE_REUSE_IMAGE`.
+
+To avoid the test framework cleaning up the image it creates, you can
+run it like this:
+
+```sh
+UACLIENT_BEHAVE_IMAGE_CLEAN=0 tox -e behave
+```
+
+which will emit a line like this above the test summary:
+
+```
+Image cleanup disabled, not deleting: behave-image-1572443113978755
+```
+
+You can then reuse that image by plugging its name into your next test
+run, like so:
+
+```sh
+UACLIENT_BEHAVE_REUSE_IMAGE=behave-image-1572443113978755 tox -e behave
```

Comment:
We don't use those flags anymore and the `reuse_image` ones will only be used if we use the `SNAPSHOT` strategy. Should we still support this ?

Reply:
We should probably remove support for this+reuse_container unless we still have a use-case. I think it might make sense to do that in a separate PR though?

Reply:
Agreed. I can do that after this PR then